### ISSUE TITLE
Alternator: support "transactions" feature

### ISSUE THREAD

**Issue by nyh** [scylladb/scylladb#5064]
"Transactions" are a new DynamoDB feature ([announced](https://aws.amazon.com/blogs/aws/new-amazon-dynamodb-transactions/) in November 2018) which allows atomic (all-or-nothing) updates to one or more tables in the same region. This feature is unrelated to the older single-item conditional updates feature - which in Scylla are called **lightweight** transactions, and which we already support after #5054.

See https://docs.google.com/document/d/1i4yjF5OSAazAY_-T8CBce9-2ykW4twx_E_Nt2zDoOVs/edit# (section "Transactions") for some implementation ideas.

The new API requests for this feature: `TransactWriteItems` and `TransactGetItems`. See [Amazon's description of this feature](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/transaction-apis.html).

**Comment by nyh** [comment]
In https://docs.google.com/document/d/1i4yjF5OSAazAY_-T8CBce9-2ykW4twx_E_Nt2zDoOVs/edit#heading=h.qmq6kez54pdq I mentioned that:

> Before this feature [transactions] was added, users were encouraged to use [Amazon’s transaction client library](https://github.com/awslabs/dynamodb-transactions) which implements transactions on top of other DynamoDB features. We can encourage them to do the same. Perhaps eventually we can take ideas from this library on how to implement transactions on the server side. 

This may be a good way to have full support for this feature - albeit somewhat ugly and slow - quickly, without huge investments in Scylla infrastructures to support transactions in CQL. Amazon have good design document https://github.com/awslabs/dynamodb-transactions/blob/master/DESIGN.md on how they implemented full-fledged multi-item cross-table transactions over pre-existing DynamoDB features, and I'm optimistic we can do the same while emulating their new transaction API. I'm optimistic that this can be done in 1-2 weeks instead of the original 1-2 month estimate (which itself was optimistic) for a from-scratch implementation of this feature.

**Comment by nyh** [comment]
I summarized the option of providing transactions over **existing** features in a new design document https://docs.google.com/document/d/1d0RmuWPkGmoKPXwJ2mY6HLhf40otuMvfOKNUYcfn2Cg/edit#. There are good news and bad news. The good news is that this I think this technique can work. The bad news is that the performance will be bad (e.g., a transactional write will be 7-9 times more expensive than a regular LWT write, while the ratio is 2 in DynamoDB), and some edge cases involving races of transactional and non-transactional operations, or interaction of transactions and GSI or streams, will be tricker to solve.

**Comment by nyh** [comment]
Our LWT implementation already supports [conditional batches](https://docs.scylladb.com/using-scylla/lwt/#batch-statements): _"The entire conditional batch has an isolated view of the database and is executed using all-or-nothing principle."_.

We could use this existing capability to implement the DynamoDB Transactions feature with one important limitation - that all operations in a transaction must be on the same partition. Although this limited feature is not as powerful as the full DynamoDB feature (that can cross partitions and even tables), it will probably be good enough for some users - and we should hopefully be able to provide this subset quickly and most importantly - efficiently (it will have the same performance as the current CAS-based writes).

**Comment by nyh** [comment]
Amazon's FAST '19 presentation by Doug Terry on how DynamoDB transactions were implemented: https://www.usenix.org/conference/fast19/presentation/terry
A nice summary of their approach and why it is simpler than other possible approaches: https://www.mydistributed.systems/2021/12/dynamodb.html

**Comment by jexh** [comment]
Is there a core scylladb issue describing the requirements to support this issue? I can't find anything that requests cross-table transaction support in the issue list - perhaps I just can't think of the correct search phrase?

**Comment by nyh** [comment]
> Is there a core scylladb issue describing the requirements to support this issue? 

I don't think there is such an issue open. I agree that it would also be nice to have multi-item transactions in CQL and that whatever core feature we'll end up implementing may also be used in CQL to implement a similar transaction feature. The reason why I opened this issue specifically for Alternator (ScyllaDB's DynamoDB-compatible API) is that it is a requirement for full DynamoDB compatibility, and that feature has known behavior (defined by Amazon). For the CQL feature it would be a completely new feature and we'll need to design it and determine how exactly it should behave. Or maybe it should be compatible with the planned Cassandra feature - https://cwiki.apache.org/confluence/display/CASSANDRA/CEP-15%3A+General+Purpose+Transactions

**Comment by jexh** [comment]
Thanks for the background. From what you say, it seems to me that it would
now be useful to open a separate issue for multi-item transaction support
in CQL. Such an issue would be a good place to capture requirements like
supporting this Alternator extension but also provide a place for
discussing the CQL specific design aspects. I think that such a feature is
very desirable in its own right and the probability of starting to really
tackle issues such as this one will be much higher if we really recognize
the benefits it could bring to the whole product.

**Comment by nyh** [comment]
Originally, DynamoDB transactions were limited to contain 10 read or write requests. Starting in [September 2022](https://aws.amazon.com/about-aws/whats-new/2022/09/amazon-dynamodb-supports-100-actions-per-transaction/), this limit was increased to 100 operations in a single transaction.

**Comment by nyh** [comment]
I started working on functional testing for this feature - see https://github.com/scylladb/scylladb/pull/22239. But don't get too happy - this is just **tests**, not the actual feature. But writing the tests before the feature if very useful for understanding the scope of the feature, its various edge cases, and will be useful as an acceptance test when we finally decide to work on this feature.

**Comment by nyh** [comment]
> Amazon's FAST '19 presentation by Doug Terry on how DynamoDB transactions were implemented: https://www.usenix.org/conference/fast19/presentation/terry A nice summary of their approach and why it is simpler than other possible approaches: https://www.mydistributed.systems/2021/12/dynamodb.html

I just discovered that Amazon also published later a full, more detailed, paper in Usenix ATC 2023 about how they implemented DynamoDB transactions. The paper is titled "Distributed Transactions at Scale in Amazon DynamoDB". See full presentation video and PDF in https://www.usenix.org/conference/atc23/presentation/idziorek

**Comment by nyh** [comment]
User story about their application that needs the multi-item transaction feature: https://github.com/scylladb/scylladb/issues/6929#issuecomment-3694651126

**Comment by kadishmal** [comment]
For us, as a FinTech company, missing support for full multi-item ACID transactions support in ScyllaDB is the major deal-breaker for migrating to ScyallyDB. We currently heavily rely on `TransacWriteItems` API in DynamoDB with 4-8 cross-table items in the single transaction request. For a payment request to succeed it is vital to have all-or-nothing behaviour across relevant tables. For an example, one transaction may include the following change requests:

1. Deduct the amount from the source account balance.
2. Add the amount to the destination account balance.
3. Create a CREDIT transaction item against the source account to leave a trace about this deduction.
4. Create a DEBIT transaction item against the destination account to leave a trace in the history about this addition.
5. Deduct the payment fee amount from the source account balance (per-transaction fee).
6. Add the payment fee amount to the destination SYSTEM account balance.
7. Create a CREDIT transaction item against the source account to leave a trace about this fee deduction.
8. Create a DEBIT transaction item against the destination SYSTEM account to leave a trace in the history about this addition.

These operations must all succeed or none at all.

I wonder if such behaviour can be accomplished using Batch Statements in LWT?

**Comment by nyh** [comment]
> I wonder if such behaviour can be accomplished using Batch Statements in LWT?

I believe the answer is no. CQL's LWT can only work inside a **single partition**. This is more powerful than just a single item (CQL "row") as DynamoDB's original conditional updates allowed before the advent of `TransactWriteItems`. But if I understand correctly, it's not powerful enough for your needs: You probably need to at least have the source and destination accounts on different partitions (you even mentioned having multiple tables).

As I noted above, it's possible to implement a full multi-item transactions feature on top of the single-item transactions that we have in both CQL and Alternator. Cassandra even had a library to do that. But such an implementation won't be efficient, and our long-term plan probably won't be do that. Rather, we are more likely to add an efficient multi-item transaction feature not built on top on top of the existing Paxos-based LWT.

**Comment by nyh** [comment]
> UPDATE: **CEP-15: General Purpose Transactions** was officially accepted in Cassandra, there is an issue - https://issues.apache.org/jira/browse/CASSANDRA-17092, and will probably miss Cassandra 5.0 and be in Cassandra 5.1. So to maintain Cassandra compatibility, the CQL feature was already defined for us.

Another update: **Cassandra 6.0 alpha1** is out since April 2026.  Its biggest feature is Accord - implementing multi-row and even multi-partition transactions. So now both Cassandra and DynamoDB have this feature and we don't have it yet.

**Comment by nyh** [comment]
Beyond `TransactWriteItems`'s ability to work on multiple items, it has another very important feature that the older single-item transactions that Alternator supports is missing: It has retry protection for non-idempotent requests - with its new **`ClientRequestToken`** parameter. Some Alternator users need this feature even for single-item transactions, and we can give it to them without inventing new APIs by implementing a `TransactWriteItems` feature limited to just a single item (or actually, LWT even allows multiple items if they are in the same partition!).

We can start with implementing this subset of `TransactWriteItems`. It will be significantly easier than implementing the full feature. All we'll need to do is to add support for storing `ClientRequestToken` of successful transactions, probably in the paxos table or a new adjacent table.

Since I haven't touched the issue of `ClientRequestToken` in this issue yet, let me spend a few paragraphs to explain it:

Imagine that you have an attribute called `balance` saving the amount of money that a customer of a bank has. Your application wants to add $1000 to a customer's record.

Your first idea is to do an UpdateItem with update expression `balance = balance + 1000`.  This is fine, and DynamoDB (and Alternator) correctly isolates concurrent balance updates. But now imagine that there was some weird error during this operation and the application has no idea if this write succeeded or didn't succeed. It's easiest to image that the application lost its network connectivity for a while, and didn't get any successful or unsuccessful response. Now, should the application retry the operation or not? If it doesn't retry, the customer might not get his $1000 at all. If it does retry the operation, and the original operation actually succeeded without the application knowing, the customer will get $2000... But how can the application know what to do? It can't. Let's say it reads the current balance and sees $6767 - this doesn't help... Does this $6767 include the $1000, or not? You can't know.

So you try to be clever and come up with a more elaborate way to add $1000: You read the current balance, find its $1234, and send an update with "set balance to $2234 if balance=$1234". But now, again, some disaster strikes and you don't get any response for this request and don't know if it succeeded or not. So you read the balance again and find there $1234. It doesn't mean the operation didn't succeed - it's also possible that it did succeed, but was immediately followed by an operation to spend this $1000! So again you don't know if you can retry! Or you see the balance is $2235 - does it mean the $1000 addition was successful together with later someone else added $1 - or maybe one other transaction added $1001 and our original addition was never successful? This get even worse if you consider the fact that an operation can be _delayed_: It's possible the original unfinished operation actually got delayed for a few seconds, but the application doesn't know because it lost the connection. So it retries and adds $1000, and a few seconds later suddenly the original operation wakes up and adds another $1000.

These "idempotency" problems **do not have a solution** in the context of either DynamoDB's classic single-item operations of ScyllaDB's LWT. The solution to them is `ClientRequestToken`, which I propose here to add:

The application that wants to add $1000 invents a random "token". Say, "k7Qm2zP9aL". It sends the transaction with `ClientRequestToken=k7Qm2zP9aL`. The application now loses the connection and has no idea if it finished or not. So it immediately creates a new connection and retries - the same operation, with the same ClientRequestToken=k7Qm2zP9aL. Now, it doesn't matter if one of these requests were lost or both were handled by the server, the server remembers it already completed a transaction with a token k7Qm2zP9aL (DynamoDB keeps the tokens of completed transactions for **10 minutes**), so when it is asked to do it a second time, it just ignores the second one. It doesn't matter if one of the operations go cancelled or not or even delayed (by not more than 10 minutes) - only one will ever happen. If the second request ends up in an unknown state too, the client is free to send it a third time. All within a 10 minutes window (which is assumed to "be enough for anybody(TM)". 

CC @ScyllaPiotr @dkropachev 

P.S. Believe it or not, Cassandra 6.0's Accord - their transaction solution - doesn't have anything like `ClientRequestToken` (a client-supplied token to allow save retries). I don't know if they just missed the need for this (!?) or for some reason they are convinced it's not really needed - perhaps because the application is expected to solve this problem itself by using a different data model - an _idempotent data model_ - or something of this sort. I can't find any public discussion on why Accord - which came after DynamoDB's implementation - doesn't have this feature.

**Comment by nyh** [comment]
After presenting above the _thesis_ that `ClientRequestToken` is an essential feature and it's shocking that neither the old DynamoDB single-item transactions or Cassandra 6.0's According had this feature, I want to also present the _anti-thesis_: That claims `ClientRequestToken` isn't really an important feature, and it doesn't really solve anything:

The example I gave above to why the `ClientRequestToken` feature is necessary assumed that we have a client application that has an intrinsic desire to add $1000 to some account's balance. It contacts the database to perform this transaction, and if for some reason this doesn't work - it wants to repeat this attempt, again and again, until the transaction is really successful. But... Is this really the case that such an application with _intrinsic_ desire exists? What if the application **itself** breaks, e.g., its machine reboots? Will its desire to add $1000 be forgotten? That doesn't sound right. So a usual solution is to have an intent log: When the application decides it wants to add $1000 to a row, the application writes down an "intent" to add $1000 to the row. If the application ever reboots, or moves to a different host, or whatever, then it wakes up and looks at its list of unfulfilled intents, sees it still has an unfulfilled intent, and performs it - asks to add $1000 to the desired account. But how do you prevent the risk of accidentally doing the $1000 addition twice, e.g., if an application reboot happens during this operation? Well, it's easy with **multi-item transactions**: The application performs a single transaction with two operations: 1. Add $1000 to the desired account, 2. Delete the intent from the log (failing if it's not there). This multi-item transaction guarantees that either both operations succeeded, or both didn't: If $1000 was successfully added, the intent is deleted and the application won't try it again. If the application wakes up and see the intent still in the log, it can retry the same two-item transaction - it's safe because the same transaction can never succeed twice because it can't delete the intent from the log twice - only one of the competing retries will succeed (and add $1000), while the rest will fail.

So according to this, if you have multi-item transactions, you can get along fine without `ClientRequestToken`, and what it provides to systems that don't have multi-item transactions (e.g., Alternator) is only partial relief.

I asked ChatGPT to provide a _synthesis_ to the thesis (that `ClientRequestToken` is essential) and the anti-thesis (that it's redundant), here is what it explained:

ClientRequestToken and the “durable intent + atomic consume” approach solve the same class of retry/ambiguity problems, but at different layers. If an application persists an intent and can execute one atomic transaction that both applies the effect (e.g., credit +$1000) and marks the intent consumed, then request-level idempotency is largely redundant for that workflow: retries won’t double-apply because only one attempt can consume the intent. In that sense, ClientRequestToken isn’t fundamental—you can build correctness without it. Where ClientRequestToken becomes “essential” in practice is when you can’t rely on those assumptions everywhere: not all operations are naturally modeled with an intent log, not all systems can cheaply do a single atomic “effect + consume” transaction, and real systems have retries at many boundaries (SDKs, proxies, timeouts, queue redelivery, partial failures). A token gives a standardized, low-friction safe-retry mechanism at the API boundary so correctness doesn’t depend on every client implementing a durable transactional state machine perfectly; it’s a guardrail, while the intent-log pattern is a stronger but more heavyweight architectural commitment.

**Comment by ScyllaPiotr** [comment]
> how do you prevent the risk of accidentally doing the $1000 addition twice, e.g., if an application reboot happens during this operation? Well, it's easy with multi-item transactions: The application performs a single transaction with two operations: 1. Add $1000 to the desired account, 2. Delete the intent from the log (failing if it's not there). This multi-item transaction guarantees that either both operations succeeded, or both didn't: If $1000 was successfully added, the intent is deleted and the application won't try it again. If the application wakes up and see the intent still in the log, it can retry the same two-item transaction - it's safe because the same transaction can never succeed twice because it can't delete the intent from the log twice - only one of the competing retries will succeed (and add $1000), while the rest will fail.

The real rationale for this working is not because of a multi-item transaction itself, but because there is now an intent log inside a client app **and** the log is used in a multi-item transaction manner.

This concept is interesting, although it silently passes over several problems:
- using intent log is another point of possible failure; that's exactly why it must be used in a multi-item transaction manner.
- the example with the client application dying or being moved to a different host - this is solving the problem of a database being not reliable by believing the client app which is using and intent log is reliable. One must ensure that only a single client app instance exists, or that the (shared) intent log is fully consistent. One must ensure that the intent log is always fully mutated (mutated successfully or that a mutation failed and did not apply, never mid-way).

So with the intent log we are essentially only passing the problem from one place to the other.

**Comment by kadishmal** [comment]
If you take a look again at my comment above in https://github.com/scylladb/scylladb/issues/5064#issuecomment-3694711785, I mentioned that we record CREDIT and DEBIT transactions **along** with deducting and adding balances.

1. The CREDIT/DEBIT transactions, located in a separate "payments" table, have unique `id`s.
2. In `TransactWriteItems` we use `ConditionExpression: 'attribute_not_exists(id)'` clause which guarantees that the CREDIT transaction item does not already exist in the table. This removes the necessity for `ClientRequestToken`.
3. If our machine/vm/container/process dies when finalizing the payment, the client retries using the unique payment ID, used as an idempotency token similar to `ClientRequestToken`. This token is used to generate the transaction ID for CREDIT/DEBIT transactions. This way there is always a guarantee that we never have duplicates. All retries must always be retried by the client. If the duplicate item is found (because the `TransactWriteItems` fails with `ConditionalCheckFailed`), we return the error and the client decides to rethrow the error or considers the job done and shows the success to the user.

Below is a simplified example. Note that `transactionId, lastTransactionDate, version` attributes are optional but useful in our business case. We use `amountNegative` because DynamoDB does not have `-` numerical operator. We need to add a negative value to reduce the balance.

```
{
  TransactItems: [
    {
      Update: {
        TableName: 'Accounts',
        Key: {
          id: 'from-account',
        },
        ConditionExpression: 'attribute_exists(id) AND balance.amount >= :amountPositive',
        UpdateExpression:
          'SET balance.amount = balance.amount + :amountNegative, transactionId = :transactionId, lastTransactionDate = :lastTransactionDate ADD version :version',
        ExpressionAttributeValues: {
          ':amountNegative': -200,
          ':amountPositive': 200,
          ':transactionId': '1636026186643_123asd_CREDIT',
          ':lastTransactionDate': 1636026186643,
          ':version': 1,
        },
      },
    },
    {
      Put: {
        TableName: 'Payments',
        Item: {
          id: '1636026186643_123asd_CREDIT',
          accountId: 'from-account',
          amount: 200,
          userId: 'from-user',
          status: 'PENDING',
          ...
          requestDate: 1636026186643,
          transactionDate: 1636026186643,
          transactionType: 'CREDIT'
        },
        ConditionExpression: 'attribute_not_exists(id)',
      },
    },
  ],
}
```

The above is a two-item transaction when a payment is in PENDING state for remote service providers. We have synchronous payments where we have four-item transactions with another item to increase the balance, and the fourth - to record the DEBIT transaction for the recipient.

I hope this example clarifies the task. Let me know if you need more explanations.
